File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int si = 0;
        int ei = nums.size()-1;
        
        if(nums[0] < nums[ei]) return nums[0];
        
        if(nums.size() == 1) return nums[0];
        
        while(si <= ei){
            int mid = si + (ei-si)/2;
            
            if(nums[mid] > nums[mid+1]){
                return nums[mid+1];
            }
            
            if(nums[mid] < nums[mid-1]){
                return nums[mid];
            }
            
            if(nums[mid] > nums[0]) si = mid+1;
            else ei = mid-1;
        }
        
        return INT_MAX;
    }
};

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums)-1
        idx = -1
        
        while left<= right:
            mid = (left+right)//2
            
            if nums[mid] <= nums[-1]:
                right = mid-1
                idx = mid
            else:
                left = mid+1
        
        
        return nums[idx]